---
title: "Identifying Differentially Expressed Genes During mtDNA Depletion and Recovery Timecourse"
author: "Jack Cazet"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

The purpose of this document is to identify differentially expressed genes across the mtDNA depletion and recovery timecourse from the raw read count matrix generated by rsem. 

```{r libraries loading, echo=FALSE}
library(edgeR)
library(knitr)
library(DT)
```

## Loading and normalizing count data

Here I set the cutoffs that are used to identify significant differences in pairwise test. Currently, I am using only an adjusted p-value cutoff of 0.01, but there is an option to add a log2 fold-change cutoff if desired.

```{r set cutoffs}
foldChange <- 0
pval <- 1e-2
```

First I load in the count matrices for the timecourses performed using different media conditions (d vs. up). I then reformat the column names to make the sample names clearer and easier to work with. The sample names had the following format: mediaCondition_timePoint_treatment (e.g., `UP_D5_DDC` refers to ddC-treated samples on depletion day 5 in UP media) I then initialize the DGElist object (analysis object used by edgeR) with the combined read counts, treatment group assignments for each sample, and gene IDs. I then perform some initial filtering to remove lowly expressed genes that are likely to be noisy/inaccuratly quantified. I used a cutoff where only genes with at least 1 count per million in at least three samples were retained for subsequent analysis.

```{r load data and initialize DGElist object}
#load read counts (genome-mapped)
counts.ns <- read.table("../../data/dge/up.mat")
counts.s <- read.table("../../data/dge/d.mat")

counts <- cbind(counts.ns,counts.s)

#rename sample names in columns to make them more readable and easier to group
colnames(counts) <- gsub('_.*','',colnames(counts))
colnames(counts) <- gsub('^(.)','\\1_',colnames(counts))
colnames(counts) <- gsub('(_UP|_D)','\\1_',colnames(counts))
colnames(counts) <- gsub('(\\d)(.+)$','\\1_\\2',colnames(counts))

#define treatment groups
TR <- factor(gsub('^[^_]*_','',colnames(counts)))

#add row with gene IDs to counts
counts$ID <- rownames(counts)

#initiate DGElist object
analysisList <- DGEList(counts = counts[,1:(ncol(counts)-1)], group = TR, genes = counts$ID)

#exclude lowly expressed genes from the analysis 
#(at least three samples have at least two counts per million)
keep <- rowSums(cpm(analysisList)>=1) >= 3
analysisList <- analysisList[keep, , keep.lib.sizes=F]

```

I then normalize the normalize the libraries and fit the model, which will serve as the basis for subsequent pairwise tests of significance. The analysis is performed across all samples in an "ANOVA-like" approach.

```{r normalize data and fit model}

#calculate normalization factors for each sample
analysisList <- calcNormFactors(analysisList)

#define treatment groups and contrasts within DGElist object (ANOVA-like setup)
design <- model.matrix(~0+TR, data = analysisList$samples)
colnames(design) <- gsub("^TR","",colnames(design))

#calculate gene-wise dispersion
analysisList <- estimateGLMRobustDisp(analysisList,design)

#fit model (basis for statistical tests)
fit <- glmQLFit(analysisList, design, robust = T)

```

## Defining pairwise contrasts and generating results tables

I next define the specific pairwise comparisons to be performed. These include comparisons within a across treatments for a given timepoint and media condition (e.g., `UP_D5` is looking for genes that are differentially expressed when comparing ddC and control samples from depletion day 5 under UP media conditions). I also compared across media conditions to identify genes that changed in significantly different ways depending on the media (e.g., `D5_UvD` compares the ddC induced changes in UP media conditions to the ddC induced changes in the D media conditions).

```{r define contrasts}

#define pairwise comparisons to be used
my.contrasts <- makeContrasts(
  #compare control and DDC contrasts within the same timepoint (UP treatment only)
  UP_D5 = `UP_D5_DDC` - `UP_D5_CTRL`,
  UP_D3 = `UP_D3_DDC` - `UP_D3_CTRL`,
  UP_R1 = `UP_R1_DDC` - `UP_R1_CTRL`,
  UP_R3 = `UP_R3_DDC` - `UP_R3_CTRL`,
  UP_R5 = `UP_R5_DDC` - `UP_R5_CTRL`,
  
  #compare control and DDC contrasts within the same timepoint (D treatment only)
  D_D5 = `D_D5_DDC` - `D_D5_CTRL`,
  D_D3 = `D_D3_DDC` - `D_D3_CTRL`,
  D_R1 = `D_R1_DDC` - `D_R1_CTRL`,
  D_R3 = `D_R3_DDC` - `D_R3_CTRL`,
  D_R5 = `D_R5_DDC` - `D_R5_CTRL`,
  
  #compare DDC-induced changes across the different media conditions
  #Which genes had a significantly different response to DDC depending on media supplementation?
  D5_UvD = (`UP_D5_DDC` - `UP_D5_CTRL`) - (`D_D5_DDC` - `D_D5_CTRL`),
  D3_UvD = (`UP_D3_DDC` - `UP_D3_CTRL`) - (`D_D3_DDC` - `D_D3_CTRL`),
  R1_UvD = (`UP_R1_DDC` - `UP_R1_CTRL`) - (`D_R1_DDC` - `D_R1_CTRL`),
  R3_UvD = (`UP_R3_DDC` - `UP_R3_CTRL`) - (`D_R3_DDC` - `D_R3_CTRL`),
  R5_UvD = (`UP_R5_DDC` - `UP_R5_CTRL`) - (`D_R5_DDC` - `D_R5_CTRL`),
  
  levels=design)

```

I then perform the pairwise comparisons defined above and export results tables. For each pairwise comparison, four tables are generated. For a comparison with the name `prefix` the tables would be: `prefix`, which contains the full results table generated by edgeR (including non-significant genes); `prefix.DG`, which contains only those genes that passed the significance cutoff; `prefix.DG.up`, which contains all significantly upregulated genes; and `prefix.DG.down`, which contains all significantly downregulated genes.

The results tables for all pairwise comparisons are stored in a list called `allRes`. Note that each results table also includes functional annotation information from (taken from uniprot).

```{r generating results tables}
genResTables <- function(x,pv,fc) {
  #pull full results of the desired pairwise contrast from the fitted model
  y <- glmQLFTest(fit, contrast = my.contrasts[,x])
  
  #extract results table from pairwise comparison
  z <- y$table
  z$ID <- rownames(z)
  z <- merge(z,Annotations, by = "ID", all.x = T)
  z$FDR <- p.adjust(z$PValue, method = "BH")
  
  #apply significance filters
  z.DG <- z[z$FDR <= pv,]
  z.DG.up <- z.DG[z.DG$logFC > fc,]
  z.DG.down <- z.DG[z.DG$logFC < -fc,]
  
  #return the four results objects as a list
  return(list(z,z.DG,z.DG.up,z.DG.down))
}

#load functional annotation table (downloaded from uniprot)
Annotations <- read.delim("../../data/generalResources/hsUpAnnots.tsv", stringsAsFactors = F)
Annotations$ID <- Annotations$Gene.Names..primary.
Annotations <- Annotations[!duplicated(Annotations$Gene.Names..primary.),]

#use the above-defined function to generate all results objects
#systematize object naming based on contrast names
allRes <- list()

for (i in 1:length(colnames(my.contrasts))) {
  
  #generate list of results objects
  resList <- genResTables(colnames(my.contrasts)[i],pval,foldChange)

  #rename objects so that they include the name of the contrast performed  
  names(resList) <- c(colnames(my.contrasts)[i],
                      paste0(colnames(my.contrasts)[i],".DG"),
                      paste0(colnames(my.contrasts)[i],".DG.up"),
                      paste0(colnames(my.contrasts)[i],".DG.down"))
  
  #add pairwise comparison results to the full results list object
  allRes[[colnames(my.contrasts)[i]]] <- resList
  
}

resultsSummary <- data.frame(comparisonName = names(allRes), 
                             upGenes = vapply(allRes, function(x) nrow(x[[3]]), numeric(1)),
                             downGenes = vapply(allRes, function(x) nrow(x[[4]]), numeric(1)))

resultsSummary$nsGenes <- nrow(allRes[[1]][[1]]) - (resultsSummary$upGenes + resultsSummary$downGenes)

datatable(resultsSummary, options = list(pageLength = length(allRes)), rownames = F)
```
## Exporting Results

Finally, I export the results for further analysis. Specifically, I export all pairwise results tables as CSVs and save the normalized counts (as log2 CPM values), the DGElist object (contains the full DGE results), and the list object of pairwise comparison tables.

```{r exporting results}
#export counts
normalizedCounts <- as.data.frame(cpm(analysisList, normalized.lib.sizes=T, log = T))

#save results dataframes, the DGE object, Annotations, and the normalized counts

# save results tables in a csv (more git friendly format)
invisible(lapply(1:length(allRes), function(x) lapply(1:4, function(y) {
  write.csv(allRes[[x]][[y]], file = paste0(names(allRes[[x]])[y],".csv"))
})))

#specify objects to be saved to binary Rdata file
saveThese <- c("allRes","analysisList","normalizedCounts")

save(file = "RNA_DGE.RData", list = saveThese)

```
